Migrate to Next.js

hakimelek opened the pull request:
This PR should add support for SSR and some other features, it should resolve #310.

It should do the following:
- [X] Replace create-react-app with Next.js.
- [X] Move page components into `pages` folder structure.
- [X] Replace all `Link` from react-router to Next.js.
- [X] Move static file from `public` to `static` folder as recommended by Next.
- [X] Main routes are rendering fine.
- [X] Redux integration.
- [X] Network dependent components are not being rendered on the server. 
- [X] Working production build.
- [x] Dynamic routes such as `/lock/:lockaddress` are working
- [x] All tests are passing.
- [x] Stories in storybook are working fine.
- [x] Add withConfig tests.

hakimelek commented on unlock-app/src/pages/creator/index.js:
```diff
@@ -0,0 +1,112 @@
+import React from 'react'
```
Previously `creator/LockMaker.js`, not sure if we should decouple this into a separate component if we would like to reuse it and load it into the page.

hakimelek commented on unlock-app/src/pages/creator/index.js:
```diff
@@ -0,0 +1,112 @@
+
+    return (
+      <Layout title="Creator">
+        <NoSSR>
```
Do not server side render this part.

hakimelek commented on unlock-app/src/pages/provider.js:
```diff
@@ -0,0 +1,69 @@
+import PropTypes from 'prop-types'
```
Previously `Web3Provider.js`.

julien51 replied:
hum, not sure I follow? This is the `/provider` page, right?

hakimelek replied:
Yes, used to be:
```
import Provider from './Web3Provider'
<Route path="/provider" component={Provider} />
```

hakimelek commented on unlock-app/src/pages/provider.js:
```diff
@@ -0,0 +1,69 @@
+const Page = withConfig(connect(mapStateToProps, mapDispatchToProps)(Web3Provider))
+
+export default (pageProps) =>
+  <NoSSR>
```
Do not server side render the entire page.

hakimelek commented on unlock-app/src/utils/withConfig.js:
```diff
@@ -1,24 +1,53 @@
+    const reduxStore = props.store
+    const router = props.router
+
+    if (!isServer) {
```
Add error state for every page wrapped in `withConfig` if the config is not loaded.

julien51 commented:
Also, I just added this issue to our Gitcoin so you can claim the bounty once it's merged ;) Please apply for it! There should be a comment right here ⬇️ in a couple minutes.

hakimelek commented on unlock-app/.babelrc:
```diff
@@ -0,0 +1,31 @@
+{
```
Nope, this is overriding .babelrc in Next.js to support styled-components as well as have control how babel should behave in different environments production, dev and test.

julien51 replied:
ok. thanks.

hakimelek commented on README.md:
````diff
@@ -40,7 +40,7 @@ $ npm run install-deps
 
 ```
-cd unlock-app && npm start
+cd unlock-app && npm run dev
````
This just starting the app in `dev` mode. `start` is kept for production mode.
👉 https://github.com/zeit/next.js/tree/b1c4f3aec4ca2d51b0ed849914cc8fc97805740a#setup

julien51 replied:
ok

hakimelek commented on unlock-app/package.json:
```diff
@@ -3,33 +3,41 @@
     "react-redux": "^5.0.7",
     "react-router-dom": "^4.2.2",
     "react-router-redux": "^5.0.0-alpha.9",
-    "react-scripts": "^2.0.4",
```
Removing `react-scripts`.

hakimelek commented on unlock-app/package.json:
```diff
@@ -3,33 +3,41 @@
     "bootstrap": "^4.0.0",
+    "color": "^0.11.0",
     "color-scheme": "^1.0.1",
+    "next": "^7.0.1",
```
Adding `next`

hakimelek commented on unlock-app/package.json:
```diff
@@ -3,33 +3,41 @@
-    "enzyme": "^3.4.0",
-    "enzyme-adapter-react-16": "^1.2.0",
-    "eslint-config-standard": "^12.0.0",
+    "babel-jest": "^23.6.0",
```
Adding `babel-jest` and `jest`. used to be a dependency of `react-scripts`.

hakimelek commented on unlock-app/package.json:
```diff
@@ -38,31 +46,33 @@
     "deploy-staging": "npm run build && TARGET=staging npm run build-unlockjs && aws s3 sync build/ s3://staging.unlock-protocol.com --profile unlock --no-progress | awk '{print $2;}' | cut -d'/' -f2- | awk '{print \"/\"$0}' | xargs aws cloudfront create-invalidation --distribution-id E1ZWU0E90B9EBN --paths --profile unlock",
-    "build": "react-scripts build",
     "build-unlockjs": "./node_modules/webpack/bin/webpack.js --config webpack-unlockjs.config.js",
-    "serve-unlockjs": "TARGET=dev node_modules/.bin/webpack-dev-server --config webpack-unlockjs.config.js &",
```
No longer needed since it's managed by `npm run dev` aka `next`.

julien51 replied:
ok this too must be possible to put in another PR?

hakimelek commented on unlock-app/package.json:
```diff
@@ -38,31 +46,33 @@
   "engineStrict": true,
   "scripts": {
-    "start-app": "react-scripts start",
+    "dev": "npm run prestart && next src",
```
For dev environment, explicitly running `npm run prestart` with `npm run dev`.  `prestart` will automatically run in production.

julien51 commented on unlock-app/src/createUnlockStore.js:
```diff
@@ -36,13 +32,12 @@ export default function createUnlockStore(config, browserHistory) {
   }, loadState())
 
   const middlewares = [
-    routerMiddleware(browserHistory),
```
Is that managed directly by Next?

hakimelek replied:
The router yes, however if you would like to have access to router in the store, then we need something like [`connected-next-router`](https://github.com/danielr18/connected-next-router). If you don’t, you `router` is injected on every page as a prop and if needed in a nested component you can use [`withRouter`](https://github.com/zeit/next.js#using-a-higher-order-component).

julien51 replied:
Ok, maybe it is cleaner this way then. On components which need to know some stuff about the path we can just wrap them into `withRouter`.

julien51 commented on unlock-app/src/createUnlockStore.js:
```diff
@@ -12,17 +12,13 @@ import locksReducer  from './reducers/locksReducer'
     locks: locksReducer,
   }
 
-  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
```
👍

julien51 commented on unlock-app/src/pages/_app.js:
```diff
@@ -0,0 +1,32 @@
+class MyApp extends App {
+  constructor (props) {
+    super(props)
+    props.reduxStore.dispatch({ type: '@@INIT' })
```
Do you mind putting that as a comment in the code so that we do not lose track of that?

julien51 commented on unlock-app/src/pages/_document.js:
```diff
@@ -0,0 +1,76 @@
+    return (
+      <html>
+        <Head>
+          <link href='https://fonts.googleapis.com/css?family=IBM+Plex+Sans:300,400,500,600,700' rel='stylesheet' />
```
*this* can definetely be in another PR, right?

julien51 commented on unlock-app/src/createUnlockStore.js:
```diff
@@ -36,13 +32,12 @@ export default function createUnlockStore(config, browserHistory) {
   const middlewares = [
-    routerMiddleware(browserHistory),
-    lockMiddleware,
+    lockMiddleware
```
Hey, we use dangling comas all over as part of our style guide, this avoid adding unneeded changes when a new items is added to an array/object.

hakimelek replied:
Will update it.

hakimelek commented:
Added support for dynamic routes, please check [`server.js`](https://github.com/unlock-protocol/unlock/pull/352/files#diff-29ead36738807d52cb31d0c8080f1e52). It checks for the route, parses the lock address slug and pass it to `getInitialProps` inside the `/lock` and `/demo` page. The page component will then have access to the query via the prop. `getInitialProps` is also called on `Link` navigation client-side. 

I cleaned up the tests and fixed storybook as well.

hakimelek commented on unlock-app/src/components/interface/buttons/Button.js:
```diff
@@ -1,71 +1,74 @@
-import { Link } from 'react-router-dom'
```
Noticed your last commit after rebase to modify Button props during rebase, I kept my changes in since it affected `Link`, let me know how you would like to keep the API here.

julien51 replied:
So, I am not sure how next-router works, but the reason I replaced the links to use `Link` from react router was to make sure we benefitted from client side nav and not fully reload the whole page when a link is clicked. 

Is this similar with next-router?

hakimelek replied:
Yes, using Link will enable client side nav but it is recommended to add the anchor tag to pass the href to it (which I added here by default here if we need to use the button for external links like the Github icons).

julien51 commented on unlock-app/package.json:
```diff
@@ -3,19 +3,25 @@
     "react-test-renderer": "^16.3.2",
-    "redux": "^4.0.1",
+    "redux": "^3.7.2",
+    "redux-devtools-extension": "^2.13.5",
```
Wouldn't that be a `devDependency` ? That said a lot of dependencies in here are probably dev... probably worth a cleanup at some point.

julien51 commented on unlock-app/package.json:
```diff
@@ -39,31 +49,40 @@
+    "start": "NODE_ENV=production node src/server.js",
     "start-ganache": "ganache-cli -b 3 -i 1984 --deterministic &",
     "deploy-lock": "cd ../smart-contracts && truffle deploy --reset",
-    "prestart": "npm run start-ganache && npm run deploy-lock && npm run serve-unlockjs",
```
Could you tell me why you removed `npm run serve-unlockjs`? 
This serves the snippet of code that is being "embedded" by 3rd party sites. See https://staging.unlock-protocol.com/unlock.js

julien51 commented on unlock-app/src/App.js:
```diff
@@ -1,55 +0,0 @@
-// Modules
```
Could you explain how this is being replaced?

hakimelek replied:
It's in [`_app.js`](https://github.com/unlock-protocol/unlock/pull/352/files/a8ec7c2b1e07739294535219ad343d71fd2b592b#diff-e0bde51797b5f02e6ca7b1ef80913342) and [`withReduxStore`](https://github.com/unlock-protocol/unlock/pull/352/files/a8ec7c2b1e07739294535219ad343d71fd2b592b#diff-084b584e69b36cca934c42ef18f5faf0)

julien51 commented on unlock-app/src/__tests__/components/Unlock.test.js:
```diff
@@ -1,57 +0,0 @@
-    }
-  })
-
-  describe('when there are no providers', () => {
```
These are important tests. Are they being moved somewhere else?

hakimelek replied:
You are right, I will move them these under testing `withConfig`.

julien51 commented on unlock-app/src/components/Unlock.js:
```diff
@@ -1,69 +0,0 @@
-import PropTypes from 'prop-types'
```
Is this moved to somewhere else? Please clarify ;)

hakimelek replied:
Yes in /pages and withConfig

julien51 commented on unlock-app/src/components/interface/buttons/Button.js:
```diff
@@ -1,71 +1,75 @@
-    {children}
-    <Label>{title}</Label>
-  </Button>
+export const LayoutButton = ({href, title, children}) => (
```
Let's keep the default `href = '#'`

benwerd replied:
```suggestion
export const LayoutButton = ({href = '#', title, children}) => (
```

julien51 commented on unlock-app/src/components/interface/buttons/Button.js:
```diff
@@ -1,71 +1,75 @@
 
-export const Button = styled(Link)`
-  background-color: ${props => props.backgroundcolor || 'var(--grey)'};
+export const Button = styled.a`
```
Since you're using regular HTML `<a>` elements I worry that this will translate into regular links which are not being used by next to do client side nav?

benwerd replied:
I had the same worry but, on reading more about Next.js, I think this is pretty neat. I really like the way it falls back - great for SEO, great for low page sizes and intelligent routing.

hakimelek replied:
@benwerd Correct! more on that [here](https://github.com/zeit/next.js#with-link).

julien51 commented on unlock-app/src/components/pages/index.js:
```diff
@@ -1,9 +0,0 @@
-import Home from './Home'
```
Why remove that? It makes import easier ;)

julien51 commented on unlock-app/src/helpers/Locks.js:
```diff
@@ -1,6 +1,7 @@
 
-const config = configure(global)
+const isServer = typeof window === 'undefined'
+const config = !isServer ? configure(global) : {}
```
I'd rather have the the `isServer` logic sit in `configure`. Could you move that there?

hakimelek replied:
Makes total sense 👍

julien51 commented on unlock-app/src/pages/dashboard.js:
```diff
@@ -35,4 +41,4 @@ const mapStateToProps = state => {
 }
 
-export default connect(mapStateToProps)(Dashboard)
+export default withConfig(connect(mapStateToProps)(Dashboard))
```
Why do you need to wrap this in `withConfig` since this code does not seem to be using the config?

hakimelek replied:
If I understood this right, dashboard is depending on connecting the Ethereum network. I am wrapping every page that uses the `network` with `withConfig` since it has the "MissingProvider" and "WrongNetwork" error logic, worth adding loading state there too if it's trying to connect to the network.

julien51 replied:
Yes you are right, the dashboard *needs* the network stuff... I see your point now. It's kind of weird that we wrap this but not actually use anything here. I'll think more about it.

julien51 commented on unlock-app/src/pages/demo.js:
```diff
@@ -0,0 +1,59 @@
+import { withConfig } from '../utils/withConfig'
+
+export class Demo extends React.Component {
+  static async getInitialProps({ req, query: { lockaddress } }) {
```
This is cool!

julien51 commented on unlock-app/src/reducers/providerReducer.js:
```diff
@@ -1,10 +1,15 @@
 import configure from '../config'
 
-const config = configure(global)
+const isServer = typeof window === 'undefined'
```
Same here ;) Moving that logic to `configure` would make sure we do not forget to add it in all files ;)

hakimelek replied:
Done

julien51 commented on unlock-app/src/stories/pages.stories.js:
```diff
@@ -1,20 +1,22 @@
+import Jobs from '../pages/jobs'
 
-storiesOf('Content pages', Pages)
+storiesOf('Content pages', module)
```
This means auto-reload won't work, right?

hakimelek replied:
Nope works fine for me.

julien51 commented:
As I am running this locally I am testing a couple things manually.
For example if I force/fake the `staging` env (by tweaking config.js`), I was hoping to see an error saying that we are using the wrong network (as my local Metamask is pointing to the main net), but I see this error:
<img width="1041" alt="screen shot 2018-10-17 at 11 10 20" src="https://user-images.githubusercontent.com/17735/47096562-72ef0c00-d1fd-11e8-85b0-c50053f65f98.png">

hakimelek commented:
@julien51 ahaa so is `unlock.js` is supposed to be the embedded script? I couldn't understand what it did and why it spins up an additional dev server.

For the `TypeError: Cannot read property 'o' of undefined` I am experiencing the same error as well in master. It checks for name of the network and it's always `1984` :/

hakimelek commented:
@benwerd every page gets SSR by default and we are wrapping client-side network dependent pages/logic with no-ssr conditions. So home/jobs/about gets always SSRed however dashboard SSR just the layout part and waits for the app to bootstrap in the client to execute the rest of the code.  While the rest of the page is blank, we can use that as an opportunity to display some skeleton or loaders for perceived performance.

asood123 commented on unlock-app/package.json:
```diff
@@ -3,19 +3,24 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@babel/core": "^7.1.2",
```
FYI, have generally found that locking in version numbers (by removing carat) saves a lot of hassle later when a random package will upgrade itself on a deployment and cause unknown issues.

hakimelek replied:
Move fast, break things!

asood123 commented:
Nice work, @hakimelek. This was not a trivial issue to take on.

julien51 commented on unlock-app/src/__tests__/utils/withConfig/__snapshots__/missingProvidersWithConfig.test.js.snap:
```diff
@@ -0,0 +1,119 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Hoooo. interesting, you're using snapshot. Normally I would ask for consistency in the tests, but since I want us to move to snapshots here too, I'll accept that and see it as a nudge to get these done in the rest of the code.

hakimelek replied:
I am glad you are planning to move to use snapshots whenever you can. The shallow and find didn't play well with this test, so I decided to go with snapshots, it's easier to write and understand.

julien51 commented on unlock-app/src/server.js:
```diff
@@ -0,0 +1,36 @@
+
+      // assigning `query` into the params means that we still
+      // get the query string passed to our application
+      const path = pathname.split('/')[1]
```
It would be cool if we found a way to avoid writing the routing logic twice.

hakimelek replied:
Any ideas how you would like to see this being abstracted?

julien51 replied:
I think we can get started with some kind of manifest file with the list of routes and how they're being handled.

hakimelek commented:
Oh wow! I am glad that this monster PR found its way to unlock 🎉

julien51 commented:
Hey @hakimelek it looks like we have a serious *size* problem where each "page" is now 2.5+MB. Any idea where that's coming from and if we can fix this easily?

hakimelek commented:
@julien51 Hmm that's interesting, can you try add [Bundle Analyzer](https://github.com/zeit/next.js/tree/canary/examples/analyze-bundles). Also is the issue in the `dev` or `prod` build?

julien51 commented:
After looking into it it looks like this is a `dev` issue only which is I guess, not as bad. I'll try to debug this!
